BJS2-15676

Caveri95 opened the pull request:
- Создана конфигурация Redis, добавлен топик WRITER
- Описан сервис для работы с достижениями
- Создан обработчик событий из Redis
- Написаны тесты

Caveri95 committed:
BJS2-15676

- Создана конфигурация Redis, добавлен топик WRITER
- Описан сервис для работы с достижениями
- Создан обработчик событий из Redis
- Написаны тесты

pablo023 commented on src/main/java/faang/school/achievement/cache/AchievementCache.java:
Вообще, это через редис делается, через хэш-мапу не подойдет для высоконагруженой системы.

Caveri95 replied:
исправил на кеш в Redis

pablo023 commented on src/main/java/faang/school/achievement/cache/AchievementCache.java:
```diff
@@ -0,0 +1,40 @@
+    @Scheduled(cron = "${cache.achievement.update-schedule}")
+    public void updateCache() {
+        log.info("Starting update of  Achievement cache");
+        this.initCache();
```
Он у тебя одно и то же будет делать, как будто не имеет это смысла. Пускай обновляет только с той инфой, которая новая после предыдущего обновления.

Caveri95 replied:
Не понял, что значит после предыдущего обновления? Я один раз при запуске в кеш поместил все, что есть в БД. В след раз мне когда обновить кеш? Опять же придется все вытаскивать из БД и смотреть, что было в кеше, а что нового появилось?

pablo023 replied:
Ну у тебя по крону вызывается метод инит, который идет, и всё из базы тащит. Нам же не нужен весь контент базы, а только то, что мы ещё не тянули. Иначе не очень оптимально - просто тащим из БД то, что итак в кэше уже есть

Caveri95 replied:
Исправил

pablo023 commented on src/main/java/faang/school/achievement/config/RedisCredentials.java:
```diff
@@ -0,0 +1,22 @@
+    private int port;
+    private Channels channels;
+
+    @Data
+    static class Channels {
+        private String achievement;
+        private String follower;
+        private String post;
+    }
```
Не надо таким заниматься, выноси в отдельные классы

pablo023 replied:
И дата тебе там не нужна, ты не используешь всего, что она генерирует.

Caveri95 replied:
окей

pablo023 commented on src/main/java/faang/school/achievement/redis/handler/WriterAchievementHandlerPost.java:
А зачем он тебе нужен, если ты по сути тут просто одну строку вызова сервиса засовываешь в методы?

Caveri95 replied:
Он реализуется, когда у нас топик приходит событие на создание поста и отрабатывает конкретная ачивка WRITER. Так понимаю, что ачивок может быть несколько на случай создания постов и для каждой нужна реализация. Вроде так

pablo023 replied:
Я имею ввиду - зачем тебе нужен метод, который просто дёргает сервис? Ты можешь по сути сам сервис и дергать, без этого метода. А то он не содержит же ничего кроме того, что ты сервис этот дергаешь.

Caveri95 committed:
BJS2-15676

- Поправлен сервис кеша на Redis
- Исправлен конфиг Redis

Caveri95 committed:
BJS2-15676

- Правки

pablo023 commented:
И конфликты надо поправить

Caveri95 committed:
BJS2-15676

- Иправлен обработчик событий, убраны лишние обертки над методами
- Исправлено обновление кеша, чтобы сохранялись только те ачивки, которых нет в кеше

Caveri95 committed:
BJS2-15676

- фикс конфликтов
